### ISSUE TITLE
pkg_resources.DefaultProvider breaks on unreadable files

### ISSUE THREAD

**Issue by llchan** [pypa/setuptools#617]
`DefaultProvider._has` only checks that a file exists, and `DefaultProvider._get` reads it without catching permission issues, so if there's an unreadable file stuff will break.  The particular case I'm looking at is reading a PKG-INFO file on some scan on import, but I think it's a general issue.

One simple solution is to use `os.access` in `_has`?  In theory we should catch exceptions on the `open` call, but this isnt something that typically races so we can probably just do the simple thing.

**Comment by llchan** [comment]
On second thought, I don't know what else calls those methods, so maybe it's safer to leave that behavior intact and catch the exception further up the chain.

The place where I'm seeing this is on pkg_resources import.  It looks like on import, it it crawls site-packages and loads versions (via `find_distributions`).  Currently one bad package anywhere in site-packages causes pkg_resources to break entirely on import, and this in turn breaks packages that import pkg_resources, effectively coupling totally unrelated packages.

Steps to reproduce:

1. Find a package that's oddly packaged and has world-unreadable files.  One example offender is [progressbar](https://pypi.python.org/pypi/progressbar).  The PKG-INFO in the tarball isnt readable to the world.  Download/extract the tarball.
2.  `python setup.py install --root /`  This is part of a larger builder, so take the `--root /` as a given.  I do this with a virtualenv, but I'm guessing it applies to system-wide installations too.
3.  The PKG-INFO in site-packages/progressbar-2.3-py2.7.egg-info will be unreadable to others
4.  `python -c 'import pkg_resources'`

For more fun, install `pytz` and try to import pytz.  Now pytz is broken simply because it imports pkg_resources, even though the issue is in a totally unrelated progressbar package.  I should add that progressbar still works, it's just the metadata that's broken.

I think what we want is for `find_distributions` to skip over packages with problems (with a warning so it's not silent).  It would also be better not to do all this work on import and instead do it when it's needed, because typically people only handle ImportErrors on import.  What do you guys think?

**Comment by jaraco** [comment]
I suspect the reason nobody has reported this before is because multi-user Python environments are rare (and also packages with restrictive permissions are rare).

> I think what we want is for find_distributions to skip over packages with problems (with a warning so it's not silent).

I'm not so sure. Fail fast is often a good strategy. What if one of the packages skipped is important - then it just pushes the error down the road. At least now the error is reported where the error occurs. It might be worthwhile suppressing the error and replacing it with a warning. The question is - which errors should we replace? Just access errors?

Also, I worry that by suppressing the error we're masking a more important problem - that the permissions defined in a package's tarball are applied to the installer's system. I suspect this issue wouldn't occur if instead you built the package into a zip distribution and then installed that. Perhaps better would be for the installers to enforce a consistent permission pattern when installing files.

>  It would also be better not to do all this work on import and instead do it when it's needed, because typically people only handle ImportErrors on import.

Agreed, behavior on import is generally a bad model. This is another fundamental issue and one that I'd prefer to solve, but I'm not sure what the compatibility implications are. I know of several use-cases that rely on `import pkg_resources` to ensure that the default working set is configured. Perhaps these can be

**Comment by llchan** [comment]
I agree that being noisy is preferable in general.  I misunderstood what it was doing on import: I thought it was some sort of preloading, in which case it should be failsafe until someone actually loads something is missing.  It looks like it's instead loading the data that is more or less final, so if that's the case, you're right, it should fail fast.

It was very counter-intuitive to have pytz and progressbar both working as expected, and upgrading setuptools broke pytz with a progressbar error despite being totally unrelated packages.  I'd like to do whatever we can to prevent unexpected coupling like that.

I think there's two issues leading to this unexpected behavior:

1. The import is doing nontrivial work that can raise all sorts of exceptions.  Looking at the pytz code, they are using `pkg_resources.resource_stream` to find an asset in pytz itself, and they want this to be a safe operation so they catch and handle ImportError.  They don't handle other exceptions---I think a very prevalent assumption is that imports only raise ImportErrors.
2. Whether or not this work happens on initialization, I also dont think that pkg_resources should crawl everything without being asked to.  Callers such as pytz are looking for something very specific, and pkg_resources doesn't need to probe everything in the universe to find it.  It should do something similar to the python import mechanism, which tries a bunch of candidate paths.

Sidenote: I should also file an issue to whoever does the setup.py install to normalize metadata permissions, so a bad tarball like the progressbar one wont install files with weird permissions.  Is that also setuptools code?

**Comment by jaraco** [comment]
> Is that also setuptools code?

Yes, in most if not all cases. Even when pip installs from a source dist, it _builds_ using setuptools, so it's conceivable that if it's fixed in setuptools, it will be fixed in all cases. It's conceivable that a package built with distutils could have the same issue if someone manually downloads the tarball and setup.py installs it without setuptools being required or present. However, in this case, setuptools can still take responsibility for correcting for the issue.

**Comment by jmroot** [comment]
An example of the kind of unintuitive problems this can cause: https://trac.macports.org/ticket/59477

Something unrelated to python failed to build because it happened to use the xattr utility which happens to be a python script, which failed to run because of one unreadable dist-info dir on the user's system (which was not even needed by xattr).

**Comment by ryandesign** [comment]
This problem continues to affect users: https://trac.macports.org/ticket/71054
